File: src/main/java/november_2024/IsAnagram.java
package november_2024;

public class IsAnagram {

    public static void main(String[] args) {
        String s = "anagram";
        String t = "nagaram";

        IsAnagram isAnagram = new IsAnagram();
        boolean anagram = isAnagram.isAnagram(s, t);
        System.out.println(anagram);
    }

    public boolean isAnagram(String s, String t) {
        if (s.length() != t.length()) {
            return false;
        }
        int[] schars = new int[26];
        int[] tchars = new int[26];

        for (int i = 0; i < s.length(); i++) {
            schars[s.charAt(i) - 'a']++;
            tchars[t.charAt(i) - 'a']++;
        }

        for (int i = 0; i < schars.length; i++) {
            if (schars[i] != tchars[i]) {
                return false;
            }
        }
        return true;
    }

}

File: python/is_anagram.py
class Solution(object):
    # O(n) time | O(26) space
    def isAnagram(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        if len(s) != len(t):
            return False

        sChars = [0] * 26
        tChars = [0] * 26

        for i in range(len(s)):
            sChars[ord(s[i]) - ord('a')] += 1
            tChars[ord(t[i]) - ord('a')] += 1
        
        for i in range(len(sChars)):
            if sChars[i] != tChars[i]:
                return False
        
        return True
    

s = "anagram"
t = "nagaram"

isAnagram = Solution()

print(isAnagram.isAnagram(s, t))
